google/cloud/spanner_v1/pool.py

```diff
@@ -138,6 +163,113 @@ def session(self, **kwargs):
+            DELETE_LONG_RUNNING_TRANSACTION_INTERVAL_SEC,
+            DELETE_LONG_RUNNING_TRANSACTION_TIMEOUT_SEC,
+        )
+
+        if (
+            not AbstractSessionPool._cleanup_task_ongoing_event.is_set()
+            and not AbstractSessionPool._cleanup_task_ongoing
+        ):
+            AbstractSessionPool._cleanup_task_ongoing_event.set()
+            if self._database.logging_enabled:
+                self._database.logger.info("95% of the session pool is exhausted")
+            background = threading.Thread(
+                target=self.deleteLongRunningTransactions,
+                args=(
+                    DELETE_LONG_RUNNING_TRANSACTION_INTERVAL_SEC,
+                    DELETE_LONG_RUNNING_TRANSACTION_TIMEOUT_SEC,
+                ),
+                daemon=True,
+                name="recycle-sessions",
+            )
+            background.start()
+
+    def stopCleaningLongRunningSessions(self):
+        """Stops background task for recycling session."""
+        AbstractSessionPool._cleanup_task_ongoing_event.clear()
+        AbstractSessionPool._cleanup_task_ongoing = False
+
+    def deleteLongRunningTransactions(
+        self, longRunningTransactionInterval_sec, longRunningTransactionTimeout_sec
+    ):
+        """Recycles sessions with long-running transactions
+        :param longRunningTransactionInterval_sec: Interval for running background task in seconds.
+        :param longRunningTransactionTimeout_sec: Timeout for recycling sessions in seconds.
+        """
+        long_running_transaction_timer = time.time()
+        transactions_closed = 0
+        while AbstractSessionPool._cleanup_task_ongoing_event.is_set():
+            is_timeout_reached = (
+                time.time() - long_running_transaction_timer
+                >= longRunningTransactionTimeout_sec
+                + longRunningTransactionInterval_sec
+            )
+
+            if is_timeout_reached and transactions_closed == 0:
+                break
+
+            AbstractSessionPool._cleanup_task_ongoing = True
+            iteration_start = time.time()
+
+            # Retrieve a list of sessions to delete
+            sessions_to_delete = [
+                session
+                for session in self._borrowed_sessions
+                if (datetime.datetime.utcnow() - session.checkout_time)
+                > datetime.timedelta(seconds=longRunningTransactionTimeout_sec)
+                and not session.long_running
+            ]
+
+            for session in sessions_to_delete:
+                self._recycle_session(session, transactions_closed)
+
+            # Calculate and sleep for the time remaining until the next iteration based on the interval.
+            iteration_elapsed = time.time() - iteration_start
+            remaining_time = longRunningTransactionInterval_sec - iteration_elapsed
+            if remaining_time > 0:
+                time.sleep(remaining_time)
```

Comment:
Can we check that this is running as a bg task and during sleep a different task is being run?